packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx

```diff
@@ -0,0 +1,263 @@
+    } = useSingleDateContext();
+
+    const ref = useForwardedRef(fwdRef, null);
+    const cellRefs = refs.calendarCellRefs;
+    const headerRef = useRef<HTMLDivElement>(null);
+    const calendarRef = useRef<HTMLTableElement>(null);
+
+    const prevValue = usePrevious(value);
+    // const prevOpen = usePrevious(isOpen);
+    // const prevHighlight = usePrevious(highlight);
+
+    const monthLabel = getFullMonthLabel(month);
+
+    /** setDisplayMonth with side effects */
+    const updateMonth = (newMonth: Date) => {
+      if (isSameUTCMonth(newMonth, month)) {
+        return;
+      }
+      setDisplayMonth(newMonth);
+      const newHighlight = getNewHighlight(highlight, month, newMonth);
+      const shouldUpdateHighlight = !isSameUTCDay(highlight, newHighlight);
+
+      if (newHighlight && shouldUpdateHighlight) {
+        setHighlight(newHighlight);
+      }
+    };
+
+    /** setHighlight with side effects */
+    const updateHighlight = (newHighlight: Date) => {
+      // change month if nextHighlight is different than `month`
+      if (!isSameUTCMonth(month, newHighlight)) {
+        setDisplayMonth(newHighlight);
+      }
+      setHighlight(newHighlight);
+      requestAnimationFrame(() => {
+        const highlightedCell = getCellWithValue(newHighlight);
+        highlightedCell?.focus();
+      });
+    };
+
+    /**
+     * If the new value is not the current month, update the month
+     */
+    useEffect(() => {
+      if (
+        value &&
+        !isSameUTCDay(value, prevValue) &&
+        !isSameUTCMonth(value, month)
+      ) {
+        setDisplayMonth(getFirstOfMonth(value));
+      }
+    }, [month, prevValue, setDisplayMonth, value]);
+
+    /** Returns the current state of the cell */
+    const getCellState = (cellDay: Date | null): CalendarCellState => {
+      if (isInRange(cellDay)) {
+        if (isSameUTCDay(cellDay, value)) {
+          return CalendarCellState.Active;
+        }
+
+        return CalendarCellState.Default;
+      }
+
+      return CalendarCellState.Disabled;
+    };
+
+    /** Called when any calendar cell is clicked */
+    const handleCalendarCellClick = (cellValue: Date) => {
+      if (!isSameUTCDay(cellValue, value)) {
+        // when the value is changed via cell,
+        // we trigger validation every time
+        handleValidation?.(cellValue);
+        setIsDirty(true);
+        // finally we update the component value
+        setValue(cellValue);
+      }
+      // and close the menu
+      closeMenu();
+    };
+
+    /** Creates a click handler for a specific cell date */
+    const cellClickHandlerForDay = (day: Date) => () => {
+      if (isInRange(day)) {
+        handleCalendarCellClick(day);
+      }
+    };
+
+    // Focus trap
+    const handleMenuKeyPress: KeyboardEventHandler<HTMLDivElement> = e => {
+      const { key } = e;
+
+      // Implementing custom focus-trap logic,
+      // since focus-trap-react focuses the first element immediately on mount
```

Comment:
non-blocking but we should consider abstracting and moving this to lib

Reply:
Agree, but that will take some time to abstract correctly. I'd suggest making a ticket for that and coming back to this later